test/monotonic_resource.cpp

```diff
@@ -14,12 +15,108 @@
+    in_buffer(
+        void* ptr,
+        void* buffer,
+        std::size_t buffer_size)
+    {
+        using ptr_t = const volatile unsigned char*;
+        return reinterpret_cast<ptr_t>(ptr) >= reinterpret_cast<ptr_t>(buffer) &&
+            reinterpret_cast<ptr_t>(ptr) < reinterpret_cast<ptr_t>(buffer) + buffer_size;
+    }
+
+    bool
+    all_alloc_in_same_block(
+        monotonic_resource& mr,
+        std::size_t bytes,
+        std::size_t align)
+    {
+        // base of the block
+        auto first = mr.allocate(align, align);
+        bool result = true;
+        for (auto allocs = (bytes - align) / align; allocs; --allocs)
+            result &= in_buffer(mr.allocate(align, align), first, bytes);
+        return result;
+    }
+
 public:
+    void
+    testGeneral()
+    {
+        // test that each block gets filled to capacity
+        // and if the growth factor is correct
+        {
+            monotonic_resource mr;
+            BOOST_TEST(all_alloc_in_same_block(mr, 1024, 1));
+            BOOST_TEST(all_alloc_in_same_block(mr, 2048, 2));
+            BOOST_TEST(all_alloc_in_same_block(mr, 4096, 1));
+            BOOST_TEST(all_alloc_in_same_block(mr, 8192, 4));
+            BOOST_TEST(all_alloc_in_same_block(mr, 16384, 1));
+            BOOST_TEST(all_alloc_in_same_block(mr, 32768, 8));
+            BOOST_TEST(all_alloc_in_same_block(mr, 65536, 1));
+        }
+        // test if each allocation is aligned correctly
+        {
+            monotonic_resource mr;
+            for (std::size_t i = 0; i < 4096; ++i)
+            {
+                const auto size = ((i * 3) % 32) + 1;
+                std::size_t next = 1;
+                for (auto mod = i % alignof(std::max_align_t);
+                    mod; mod >>= 1, next <<= 1);
+                const auto align = (std::max)(next,
+                    std::size_t(1));
+                BOOST_TEST(!(reinterpret_cast<std::uintptr_t>(mr.allocate(size, align)) % align));
+            }
+        }
+        // test if user provided sizes are correctly rounded
+        {
+            {
+                monotonic_resource mr(10);
+                BOOST_TEST(all_alloc_in_same_block(mr, 1024, 1));
+            }
+            {
+                monotonic_resource mr(1025);
+                BOOST_TEST(all_alloc_in_same_block(mr, 2048, 1));
+            }
+            {
+                monotonic_resource mr(4000);
+                BOOST_TEST(all_alloc_in_same_block(mr, 4096, 1));
+            }
+        }
+        // test if sizes are correctly determined from initial buffers
+        {
+            {   
+                std::unique_ptr<unsigned char[]> buf(new unsigned char[512]);
+                monotonic_resource mr(buf.get(), 512);
+                BOOST_TEST(all_alloc_in_same_block(mr, 512, 1));
+                BOOST_TEST(all_alloc_in_same_block(mr, 1024, 1));
+            }
+            {   
+                std::unique_ptr<unsigned char[]> buf(new unsigned char[2048]);
```

Comment:
why not make this a local array, e.g.
```
char buf[2048];
```
? That's going to be the most common use-case.

Reply:
I don't think it really matters here... I can add test for both.

Reply:
its just that now we're instantiating `unique_ptr` needlessly